File: packages/agent-runtime/py/src/steerable_agent_runtime/todo.py
"""Session task list — the ``todo_write`` tool.

The model maintains one ordered task list per chat (Claude Code
``TodoWrite`` parity): multi-step work stays legible across rounds because
the model rewrites the list as it progresses, and every rewrite lands in the
durable message record as the tool result, so a resumed session rebuilds the
list from the transcript.

Semantics are full-replace: each call carries the complete list, not a diff.
The store keys lists by ``chat_id`` from the dispatch context, so concurrent
chats in one sidecar process never see each other's tasks.
"""

from __future__ import annotations

from typing import Any

from .errors import ToolDispatchError
from .hooks import CompletionAction, CompletionDraft, NoopHooks

#: The tool name the model calls.
TODO_TOOL_NAME = "todo_write"

#: Valid task states, in lifecycle order.
TODO_STATUSES = ("pending", "in_progress", "completed")

#: Bounds on the list itself. The list is a working set, not a backlog —
#: past ~50 items the model is planning, not tracking, and the rendered
#: result starts to lean on the fragment budget.
_MIN_TODOS = 1
_MAX_TODOS = 50

#: Model-facing description, shared by the registration metadata and the
#: OpenAI descriptor.
_DESCRIPTION = (
    "Track multi-step work as an ordered task list. Create the list when a "
    "task needs several steps; rewrite it as you progress — mark the task "
    "you are working on in_progress (at most one) and completed the moment "
    "it is done. Each call replaces the whole list. Skip it for single-step "
    "requests."
)

#: JSON Schema for the tool's arguments — one full-replace list.
TODO_SCHEMA: dict[str, Any] = {
    "type": "object",
    "properties": {
        "todos": {
            "type": "array",
            "minItems": _MIN_TODOS,
            "maxItems": _MAX_TODOS,
            "items": {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "string",
                        "description": "Stable identifier for this task.",
                    },
                    "content": {
                        "type": "string",
                        "description": "One line describing the task.",
                    },
                    "status": {
                        "type": "string",
                        "enum": list(TODO_STATUSES),
                    },
                },
                "required": ["id", "content", "status"],
            },
            "description": (
                "The complete task list, replacing the previous one. "
                "Mark exactly the task you are working on as in_progress; "
                "keep at most one in_progress at a time."
            ),
        },
    },
    "required": ["todos"],
}


class TodoStore:
    """Process-local task lists keyed by chat id.

    Persistence comes for free from the message record — every rewrite is a
    tool result in the transcript, and full-replace semantics make the state
    self-healing: after a restart the model's next call rewrites the whole
    list from what it sees in the resumed history.
    """

    def __init__(self) -> None:
        self._by_chat: dict[str, list[dict[str, Any]]] = {}

    def get(self, chat_id: str) -> list[dict[str, Any]]:
        return [dict(item) for item in self._by_chat.get(chat_id, [])]

    def set(self, chat_id: str, todos: list[dict[str, Any]]) -> None:
        self._by_chat[chat_id] = [dict(item) for item in todos]


def _normalize_todos(todos: Any) -> list[dict[str, Any]]:
    """Validate one full-replace list and return it in canonical shape.

    A violation raises ``ToolDispatchError`` — the router wraps it as the
    tool result, so the model sees exactly what to fix and retries within
    the same turn.
    """
    if not isinstance(todos, list):
        raise ToolDispatchError(
            f"todo_write: todos must be an array, got {type(todos).__name__}"
        )
    if not (_MIN_TODOS <= len(todos) <= _MAX_TODOS):
        raise ToolDispatchError(
            f"todo_write: todos must contain {_MIN_TODOS}-{_MAX_TODOS} items, "
            f"got {len(todos)}. Track the current working set, not a backlog."
        )
    seen_ids: set[str] = set()
    in_progress = 0
    normalized: list[dict[str, Any]] = []
    for index, item in enumerate(todos):
        if not isinstance(item, dict):
            raise ToolDispatchError(
                f"todo_write: todos[{index}] must be an object, "
                f"got {type(item).__name__}"
            )
        task_id = item.get("id")
        if not isinstance(task_id, str) or not task_id:
            raise ToolDispatchError(
                f"todo_write: todos[{index}] is missing a non-empty string "
                '"id".'
            )
        if task_id in seen_ids:
            raise ToolDispatchError(
                f'todo_write: duplicate id "{task_id}" — ids must be unique.'
            )
        seen_ids.add(task_id)
        content = item.get("content")
        if not isinstance(content, str) or not content.strip():
            raise ToolDispatchError(
                f'todo_write: todos[{index}] ("{task_id}") is missing a '
                'non-empty string "content".'
            )
        status = item.get("status")
        if status not in TODO_STATUSES:
            raise ToolDispatchError(
                f'todo_write: todos[{index}] ("{task_id}") has invalid status '
                f"{status!r}; expected one of {', '.join(TODO_STATUSES)}."
            )
        if status == "in_progress":
            in_progress += 1
        normalized.append({"id": task_id, "content": content, "status": status})
    if in_progress > 1:
        raise ToolDispatchError(
            f"todo_write: {in_progress} tasks are in_progress; keep at most "
            "one — mark the rest pending or completed."
        )
    return normalized


def make_todo_write_tool(store: TodoStore) -> Any:
    """Build the ``todo_write`` tool handler bound to ``store``.

    The returned coroutine function carries the registration metadata the
    router reads (name, description, schema) so a host registers it with one
    call. Marked ``concurrency_safe=False``: two rewrites of the same chat's
    list would race, and last-writer-wins is the wrong resolution for a
    full-replace update.
    """

    async def todo_write(
        todos: list[dict[str, Any]], context: dict[str, Any] | None = None
    ) -> dict[str, Any]:
        normalized = _normalize_todos(todos)
        chat_id = str((context or {}).get("chat_id") or "")
        store.set(chat_id, normalized)
        counts = {status: 0 for status in TODO_STATUSES}
        for item in normalized:
            counts[item["status"]] += 1
        return {
            "todos": normalized,
            "summary": {
                "total": len(normalized),
                "pending": counts["pending"],
                "inProgress": counts["in_progress"],
                "completed": counts["completed"],
            },
        }

    todo_write.__steerable_tool_meta__ = {  # type: ignore[attr-defined]
        "name": TODO_TOOL_NAME,
        "mode": "read",  # session state only; no workspace side effects
        "description": _DESCRIPTION,
        "schema": TODO_SCHEMA,
        "require_consent": False,
        "concurrency_safe": False,
        "exposure": "direct",
    }
    return todo_write


def todo_write_tool_descriptor() -> dict[str, Any]:
    """OpenAI tool schema to append to the model's tools list.

    The sidecar registers ``todo_write`` on its router (so dispatch works on
    both the host path and the sidecar-local path) but the model only sees it
    when the descriptor is appended to the ``tools`` array — mirroring how
    run_code advertises itself.
    """
    return {
        "type": "function",
        "function": {
            "name": TODO_TOOL_NAME,
            "description": _DESCRIPTION,
            "parameters": TODO_SCHEMA,
        },
    }


#: Unfinished states the completion gate acts on. ``completed`` items never
#: block finishing; a list that is all-completed is a finished plan.
_UNFINISHED_STATUSES = ("pending", "in_progress")

#: How many unfinished items the reminder names. Past this the listing stops
#: being a reminder and starts being a context dump; the count line still
#: carries the full total.
_MAX_LISTED = 10


class TodoCompletionGate(NoopHooks):
    """``before_completion`` hook: veto ``completed`` while todos remain.

    Long tasks stall when the model ends a turn with a text-only progress
    note ("next I will do page 14") — the loop reads that as ``completed``
    and waits for user input even though the chat's todo list still has
    pending or in_progress items. The gate converts that completion into a
    retry with a reminder naming the unfinished items, so the loop keeps
    executing instead of stalling.

    Bounded by the loop's completion-redo budget: a model that deliberately
    leaves items unfinished (after explaining why) is still allowed to
    finish once the budget is spent, and the loop discloses that acceptance.

    Stateless across chats: the run's ``LoopContext.chat_id`` keys the store
    lookup, matching the id ``todo_write`` dispatches with, so one gate
    instance serves every chat the process hosts.
    """

    def __init__(self, store: TodoStore) -> None:
        self._store = store

    async def before_completion(
        self, draft: CompletionDraft, ctx: Any
    ) -> CompletionAction:
        """Retry when the run's chat has unfinished todos, else accept.

        Only ``completed`` drafts are gated — ``budget_exhausted`` and error
        stops have their own continuation paths, and vetoing them here would
        fight those.
        """
        if draft.status != "completed":
            return CompletionAction(kind="accept")
        unfinished = [
            item
            for item in self._store.get(str(getattr(ctx, "chat_id", None) or ""))
            if item["status"] in _UNFINISHED_STATUSES
        ]
        if not unfinished:
            return CompletionAction(kind="accept")
        listing = "\n".join(
            f"- [{item['status']}] {item['content']}"
            for item in unfinished[:_MAX_LISTED]
        )
        return CompletionAction(
            kind="retry",
            message=(
                f"The todo list still has {len(unfinished)} unfinished "
                f"item(s):\n{listing}\n"
                "Keep executing them with tools instead of narrating "
                "progress. If an item is genuinely done or no longer needed, "
                "update the list with todo_write first, then finish."
            ),
        )

File: packages/agent-runtime/rs/src/todo.rs
//! Session task list — `todo_write` (Python `todo.py`).

use std::collections::{HashMap, HashSet};

use serde_json::{json, Map, Value};

use crate::types::ToolResult;

pub const TODO_TOOL_NAME: &str = "todo_write";
pub const TODO_STATUSES: [&str; 3] = ["pending", "in_progress", "completed"];
pub const TODO_DESCRIPTION: &str = "Track multi-step work as an ordered task list. Create the list when a task needs several steps; rewrite it as you progress — mark the task you are working on in_progress (at most one) and completed the moment it is done. Each call replaces the whole list. Skip it for single-step requests.";
const MIN_TODOS: usize = 1;
const MAX_TODOS: usize = 50;

#[derive(Clone, Debug, Default)]
pub struct TodoStore {
    by_chat: HashMap<String, Vec<Value>>,
}

impl TodoStore {
    pub fn get(&self, chat_id: &str) -> Vec<Value> {
        self.by_chat.get(chat_id).cloned().unwrap_or_default()
    }

    pub fn set(&mut self, chat_id: &str, todos: Vec<Value>) {
        self.by_chat.insert(chat_id.to_string(), todos);
    }
}

pub fn apply_todo_write(
    store: &mut TodoStore,
    args: &Value,
    chat_id: &str,
) -> Result<Value, String> {
    let todos = args.get("todos").cloned().unwrap_or(Value::Null);
    let normalized = normalize_todos(&todos)?;
    store.set(chat_id, normalized.clone());
    let mut pending = 0;
    let mut in_progress = 0;
    let mut completed = 0;
    for item in &normalized {
        match item.get("status").and_then(Value::as_str) {
            Some("pending") => pending += 1,
            Some("in_progress") => in_progress += 1,
            Some("completed") => completed += 1,
            _ => {}
        }
    }
    Ok(json!({
        "todos": normalized,
        "summary": {
            "total": normalized.len(),
            "pending": pending,
            "inProgress": in_progress,
            "completed": completed,
        }
    }))
}

pub fn todo_write_result(store: &mut TodoStore, args: &Value, chat_id: &str) -> ToolResult {
    match apply_todo_write(store, args, chat_id) {
        Ok(data) => ToolResult::ok(data),
        Err(error) => ToolResult::fail(error),
    }
}

pub fn todo_write_tool_descriptor() -> Value {
    json!({
        "type": "function",
        "function": {
            "name": TODO_TOOL_NAME,
            "description": TODO_DESCRIPTION,
            "parameters": {
                "type": "object",
                "properties": {
                    "todos": {
                        "type": "array",
                        "minItems": MIN_TODOS,
                        "maxItems": MAX_TODOS,
                        "items": {
                            "type": "object",
                            "properties": {
                                "id": {"type": "string", "description": "Stable identifier for this task."},
                                "content": {"type": "string", "description": "One line describing the task."},
                                "status": {"type": "string", "enum": TODO_STATUSES},
                            },
                            "required": ["id", "content", "status"],
                        },
                    },
                },
                "required": ["todos"],
                "additionalProperties": false,
            },
        },
    })
}

fn normalize_todos(todos: &Value) -> Result<Vec<Value>, String> {
    let Some(items) = todos.as_array() else {
        return Err(format!(
            "todo_write: todos must be an array, got {}",
            json_type(todos)
        ));
    };
    if !(MIN_TODOS..=MAX_TODOS).contains(&items.len()) {
        return Err(format!(
            "todo_write: todos must contain {MIN_TODOS}-{MAX_TODOS} items, got {}. Track the current working set, not a backlog.",
            items.len()
        ));
    }
    let mut seen_ids = HashSet::new();
    let mut in_progress = 0;
    let mut normalized = Vec::new();
    for (index, item) in items.iter().enumerate() {
        let Some(obj) = item.as_object() else {
            return Err(format!(
                "todo_write: todos[{index}] must be an object, got {}",
                json_type(item)
            ));
        };
        let task_id = obj.get("id").and_then(Value::as_str).unwrap_or("");
        if task_id.is_empty() {
            return Err(format!(
                "todo_write: todos[{index}] is missing a non-empty string \"id\"."
            ));
        }
        if !seen_ids.insert(task_id.to_string()) {
            return Err(format!(
                "todo_write: duplicate id \"{task_id}\" — ids must be unique."
            ));
        }
        let content = obj.get("content").and_then(Value::as_str).unwrap_or("");
        if content.trim().is_empty() {
            return Err(format!(
                "todo_write: todos[{index}] (\"{task_id}\") is missing a non-empty string \"content\"."
            ));
        }
        let status = obj.get("status").and_then(Value::as_str).unwrap_or("");
        if !TODO_STATUSES.contains(&status) {
            return Err(format!(
                "todo_write: todos[{index}] (\"{task_id}\") has invalid status {status:?}; expected one of {}.",
                TODO_STATUSES.join(", ")
            ));
        }
        if status == "in_progress" {
            in_progress += 1;
        }
        let mut row = Map::new();
        row.insert("id".into(), json!(task_id));
        row.insert("content".into(), json!(content));
        row.insert("status".into(), json!(status));
        normalized.push(Value::Object(row));
    }
    if in_progress > 1 {
        return Err(format!(
            "todo_write: {in_progress} tasks are in_progress; keep at most one — mark the rest pending or completed."
        ));
    }
    Ok(normalized)
}

fn json_type(value: &Value) -> &'static str {
    match value {
        Value::Null => "None",
        Value::Bool(_) => "bool",
        Value::Number(_) => "number",
        Value::String(_) => "str",
        Value::Array(_) => "list",
        Value::Object(_) => "dict",
    }
}
